source/particle/generator/reference_cell.cc

```diff
@@ -0,0 +1,192 @@
+                        const Particle<dim> particle(position_real, *itr_particles_in_unit_cell, particle_index);
+                        const types::LevelInd cell_index(cell->level(), cell->index());
+                        particles.insert(std::make_pair(cell_index, particle));
+                        particle_index++;
+                      }
+                    catch (typename Mapping<dim>::ExcTransformationFailed &)
+                      {
+                        AssertThrow (true,
+                                     ExcMessage("Couldn't generate particle (unusual cell shape?). "));
+                      }
+                  }
+              }
+          }
+
+        starting_particle_index = Utilities::MPI::max(particle_index, this->get_mpi_communicator());
+      }
+
+
+      template <int dim>
+      const std::vector<Point<dim>>
+                                 ReferenceCell<dim>::generate_particle_positions_in_unit_cell()
+      {
+        std::vector<Point<dim>> particle_positions;
+        std_cxx11::array<double, dim> spacing;
+
+        // Calculate separation of particles
+        for (unsigned int i = 0; i < dim; ++i)
+          spacing[i] = 1.0 / number_of_particles[i];
+
+        for (unsigned int i = 0; i < number_of_particles[0]; ++i)
+          {
+            for (unsigned int j = 0; j < number_of_particles[1]; ++j)
+              {
+                if (dim == 2)
+                  {
+                    Point<dim> position_unit = Point<dim>(i * spacing[0] + spacing[0] / 2,
+                                                          j * spacing[1] + spacing[1] / 2);
+                    particle_positions.push_back(position_unit);
+                  }
+                else if (dim == 3)
+                  {
+                    for (unsigned int k = 0; k < number_of_particles[2]; ++k)
+                      {
+                        Point<dim> position_unit = Point<dim>(i * spacing[0] + spacing[0] / 2,
+                                                              j * spacing[1] + spacing[1] / 2,
+                                                              k * spacing[2] + spacing[2] / 2);
+                        particle_positions.push_back(position_unit);
+                      }
+                  }
+                else
+                  ExcNotImplemented();
+              }
+          }
+
+        return particle_positions;
+      }
+
+
+      template <int dim>
+      void
+      ReferenceCell<dim>::declare_parameters (ParameterHandler &prm)
+      {
+        prm.enter_subsection("Postprocess");
+        {
+          prm.enter_subsection("Tracers");
+          {
+            prm.enter_subsection("Generator");
+            {
+              prm.enter_subsection("Reference cell");
+              {
+                prm.declare_entry ("Number of tracers per cell per direction", "10",
```

Comment:
From looking at the code I was thinking one would be able to prescribe different numbers of particle per direction. I do not know if anybody would want to do this, but in case you would be interested: There is a function called `Utilities::possibly_extend_from_1_to_N` that would allow you to either enter one number (for the same number of particles in every direction), or a list of `dim` entries for the different directions. It is for example used for the `densities` parameter of the `DiffusionDislocation` material model. If you are not interested, you can also keep everything as is (it just looked like you prepared the code, and then did not make the input parameter for different numbers).